### PR TITLE
RC-20249 Tf aks v3

### PR DESCRIPTION
**Pull request by poornima-rafay**
https://rafaysystems.atlassian.net/browse/RC-20249

Testing:
Day 0 operation: Create cluster using terraform apply - DONE
Day 2 operation: Add nodepool in spec -> terraform apply - DONE 
Delete Cluster -> terraform destroy - DONE

### CODE REVIEW

File: examples/resources/rafay_aks_cluster_v3/resource.tf [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    name    = "aks-v3-tf-2401202303"
+    project = "defaultproject"
+  }
+  spec {
+    type          = "aks"
+    blueprint_config {
+      name = "default-aks"
+      version = "1.21.0"
+    }
+    cloud_credentials = "poornima-aks-cred-dev"
```
Comment (sbenny) [inline examples/resources/rafay_aks_cluster_v3/resource.tf:12]: can you change the names in the example to more generic ones? Customers see this in the doc. Refer other examples

File: examples/resources/rafay_aks_cluster_v3/testmodule.sh [hunk 2]
```diff
@@ -0,0 +1 @@
+rm -rf .terraform* && rm -rf terraform.tfstate && terraform init && terraform init && terraform plan
```
Comment (sbenny) [inline examples/resources/rafay_aks_cluster_v3/testmodule.sh:1]: Not sure we want to have the shell script in the example folder, This is an external-facing repo. Move this to tools folder

File: examples/resources/rafay_aks_cluster_v3/variables.tf [hunk 3]
```diff
@@ -0,0 +1,5 @@
+variable "rafay_config_file" {
+  description = "rafay provider config file for authentication"
+  sensitive   = true
+  default     = "/Users/poornima.v/.rafay/cli/config.json"
```
Comment (sbenny) [inline examples/resources/rafay_aks_cluster_v3/variables.tf:4]: Please correct /Users/poornima.v... Refer to other examples

File: rafay/utils.go [hunk 4]
```diff
@@ -654,6 +655,37 @@ func expandProjectMeta(p []interface{}) []*commonpb.ProjectMeta {
...
+
+		if v, ok := in["name"].(string); ok && len(v) > 0 {
+			obj.Name = v
+			sortByName = append(sortByName, v)
+		}
+
+		out[i] = &obj
+	}
+
+	var sortedOut []*infrapb.Projects
```
Comment (sbenny) [inline rafay/utils.go:676]: Does BE gives project in sorted order, then this looks good

File: examples/resources/rafay_aks_cluster_v3/testmodule.sh (hunk 2 above)
Comment (poornima-rafay) [reply examples/resources/rafay_aks_cluster_v3/testmodule.sh]: removed the file.

File: rafay/resource_aks_cluster_v3.go [hunk 5]
```diff
@@ -0,0 +1,3116 @@
...
+			obj.Type = v
+		}
+
+		if v, ok := in["location"].(string); ok && len(v) > 0 {
+			obj.Location = v
+		}
+		outToSort[i] = obj
+	}
+
+	sort.Sort(ByNodepoolNameV3(outToSort))
```
Comment (kutumba-rafay) [inline rafay/resource_aks_cluster_v3.go:1165]: @sbenny @poornima-rafay this sorting is becoming a problem in TF as TF thinks there is a change in the state. Is there a way to preserve the order that user gives?

File: rafay/resource_aks_cluster_v3.go [hunk 6]
```diff
@@ -0,0 +1,3116 @@
...
+		obj = p[0].(map[string]interface{})
+	}
+
+	if len(in.Type) > 0 {
+		obj["type"] = in.Type
+	}
+
+	// TODO: Test this
+	if in.UserAssignedIdentities != nil && len(in.UserAssignedIdentities) > 0 {
+		obj["user_assigned_identities"] = in.UserAssignedIdentities
```
Comment (kutumba-rafay) [inline rafay/resource_aks_cluster_v3.go:1827]: Don't we need a flatten function for UserAssignedIdentities?

File: rafay/resource_aks_cluster_v3.go [hunk 7]
```diff
@@ -0,0 +1,3116 @@
...
+	if in == nil {
+		return nil
+	}
+	obj := map[string]interface{}{}
+	if len(p) != 0 && p[0] != nil {
+		obj = p[0].(map[string]interface{})
+	}
+
+	if in.AuthorizedIPRanges != nil && len(in.AuthorizedIPRanges) > 0 {
+		obj["authorized_ipr_ranges"] = toArrayInterface(in.AuthorizedIPRanges)
```
Comment (kutumba-rafay) [inline rafay/resource_aks_cluster_v3.go:2018]: same comment as in expand function. Is this (authorized_ipr_ranges) a typo?

**Review by kutumba-rafay — COMMENTED** [review]
In some places, we are not checking if the field exists and assigning it. Shouldn't we check before assigning it? Ex: 	"obj["net_netfilter_nf_conntrack_buckets"] = in.NetNetfilterNfConntrackBuckets". There are many like this.

File: examples/resources/rafay_aks_cluster_v3/providers.tf [hunk 8]
```diff
@@ -0,0 +1,12 @@
+terraform {
+  required_providers {
+    rafay = {
+      "source": "rafay/rafay",
```
[the remark this answers is not in the document]
Comment (poornima-rafay) [reply examples/resources/rafay_aks_cluster_v3/providers.tf]: Done

File: rafay/resource_aks_cluster_v3.go (hunk 5 above)
Comment (poornima-rafay) [reply rafay/resource_aks_cluster_v3.go:1165]: @kutumba-rafay I encountered this issue today. 
I am not sure how this is addressed in other tf yet. Probably need a solution at the DB level where data is stored and retrieved from in the same order. 
@sbenny please suggest whats the approach to take?

**Review by stephan-rafay — APPROVED** [review]

**Pull request merged**